lib/dynamoid/criteria/chain.rb

```diff
@@ -258,6 +258,8 @@ def range_query
       end
 
       def type_cast_condition_parameter(key, value)
+        return value if [:array, :set].include?(source.attributes[key.to_sym][:type])
```

Comment:
Shouldn't this be the `else` case below? So really the `else` case is wrong because the `dump` field won't work correctly since the type is array even though it's acting on individual values of the array.

Just quick glance, almost seems like this really should just always be, right?

```
source.dump_field(value, source.attributes[key.to_sym]
```

Reply:
@richardhsu I think it shouldn't

The value may be collection (Set or Array) in following cases:
* `in` operator - `where("field.in" => [1, 2, 3])`
* `between` operator - `where("field.between" => [0, 10])`
* `equal` operator for `set`/`array` - `where(set_field: Set.new([1, 2, 3]))`

We have to handle cases 1, 2 and 3 in different way. For cases 1, 2 we have to convert each element of collection. For case 3  we have convert the whole value (to `set` or `array`)

`else` case handles cases 1, 2. And here I disable/postpone handling case 3 in order to simplify code. It's a complex logic because we have to take into account not only field type but operator. For instance for `set` field operator `equal` expects `set` but `contains` expects type of `set` element - `number` or `string`

Reply:
Thanks for the explanation!

Reply:
Ah perfect, @andrykonchin thanks for explanation!